src/core/algorithms/sfd/sfd_algorithm.cpp

```diff
@@ -0,0 +1,302 @@
+    long double chi_squared =
+            CalculateChiSquared(n_i_j, n_i, n_j, col_i, col_k, smp.GetRowIndices().size());
+
+    long double v = (domains_[col_i] - 1) * (domains_[col_k] - 1);
+
+    boost::math::chi_squared dist(v);
+
+    long double t = quantile(dist, 1 - max_false_positive_probability_);
+    return chi_squared > t;
+}
+
+void SFDAlgorithm::CheckForCorrelation(model::ColumnIndex col_i, model::ColumnIndex col_k,
+                                       std::vector<model::TypedColumnData> const &data, Sample &smp,
+                                       FrequencyHandler const &handler) {
+    SkewHandling(col_i, col_k, data, smp, handler);
+
+    std::vector<std::vector<long double>> n_i_j(domains_[col_i],
+                                                std::vector<long double>(domains_[col_k], 0));
+    std::vector<long double> n_i(domains_[col_i], 0);
+    std::vector<long double> n_j(domains_[col_k], 0);
+
+    FillContingencyTable(smp, data, col_i, col_k, n_i_j, n_i, n_j, handler);
+
+    if (TooMuchStructuralZeros(n_i_j, col_i, col_k) ||
+        ChiSquaredTest(col_i, col_k, smp, n_i_j, n_i, n_j)) {
+        correlations_.emplace_back(col_i, col_k);
+        return;
+    }
+    unrelated_.emplace_back(col_i, col_k);
+}
+
+bool SFDAlgorithm::IsTrivialCase(model::ColumnIndex col_ind, std::vector<bool> &is_soft_or_trivial,
+                                 FrequencyHandler const &handler, size_t row_count) {
+    if (is_soft_or_trivial[col_ind]) {
+        return true;
+    }
+    if (handler.GetColumnCardinality(col_ind) >= (1 - minimum_cardinality_) * row_count) {
+        soft_keys_.push_back(col_ind);
+        is_soft_or_trivial[col_ind] = true;
+        return true;
+    }
+    if (handler.GetColumnCardinality(col_ind) == 1) {
+        trivial_columns_.push_back(col_ind);
+        is_soft_or_trivial[col_ind] = true;
+        return true;
+    }
+    return false;
+}
+
+unsigned long long SFDAlgorithm::ExecuteInternal() {
+    std::vector<model::TypedColumnData> const &data = typed_relation_->GetColumnData();
+
+    size_t row_count = data.front().GetNumRows();
+    model::ColumnIndex column_count = data.size();
+    std::vector<bool> is_soft_or_trivial(column_count, false);
+    Init(column_count);
+
+    auto start_time = std::chrono::high_resolution_clock::now();
+    FrequencyHandler handler(data, column_count, max_amount_of_categories_);
+
+    for (model::ColumnIndex ind1 = 0; ind1 < column_count - 1; ind1++) {
+        if (IsTrivialCase(ind1, is_soft_or_trivial, handler, row_count)) continue;
+
+        for (model::ColumnIndex ind2 = ind1 + 1; ind2 < column_count; ind2++) {
+            if (IsTrivialCase(ind2, is_soft_or_trivial, handler, row_count)) continue;
+
+            auto [col_i, col_k] =
+                    SwitchIndicesIfNecessary(ind1, ind2, handler.GetColumnCardinality(ind1),
+                                             handler.GetColumnCardinality(ind2));
```

Comment:
let's also make code more readable (function name `SwitchIndicesIfNecessary` is not intuitive):
```cpp
    auto sort_indices_by_cardinality = [&handler](model::ColumnIndex ind1,
                                                  model::ColumnIndex ind2) -> ColumnIndexPair {
        if (handler.GetColumnCardinality(ind2) > handler.GetColumnCardinality(ind1)) {
            return {ind2, ind1};
        }
        return {ind1, ind2};
    };
    ...

    for (model::ColumnIndex ind1 = 0; ind1 < column_count - 1; ind1++) {
        ...

        for (model::ColumnIndex ind2 = ind1 + 1; ind2 < column_count; ind2++) {
            ...
            auto [col_i, col_k] = sort_indices_by_cardinality(ind1, ind2);
```
And we use `SwitchIndicesIfNecessary`  only here, so we can define this function as lambda right here.